src/net_processing.cpp

```diff
@@ -1546,56 +1549,57 @@ PeerRef PeerManagerImpl::RemovePeer(NodeId id)
     return ret;
 }
 
-bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) const
+CNodeStateStats PeerManagerImpl::GetNodeStateStats(NodeId nodeid) const
 {
+    CNodeStateStats stats;
     {
         LOCK(cs_main);
         const CNodeState* state = State(nodeid);
-        if (state == nullptr)
-            return false;
-        stats.nSyncHeight = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
-        stats.nCommonHeight = state->pindexLastCommonBlock ? state->pindexLastCommonBlock->nHeight : -1;
+        if (state == nullptr) {
+            return stats;
+        }
+        if (state->pindexBestKnownBlock) {
+            stats.m_sync_height = state->pindexBestKnownBlock->nHeight;
+        }
+        if (state->pindexLastCommonBlock) {
+            stats.m_common_height = state->pindexLastCommonBlock->nHeight;
+        }
         for (const QueuedBlock& queue : state->vBlocksInFlight) {
-            if (queue.pindex)
-                stats.vHeightInFlight.push_back(queue.pindex->nHeight);
+            if (queue.pindex) {
+                stats.m_height_in_flight.push_back(queue.pindex->nHeight);
+            }
         }
     }
-
-    PeerRef peer = GetPeerRef(nodeid);
-    if (peer == nullptr) return false;
-    stats.their_services = peer->m_their_services;
-    stats.m_starting_height = peer->m_starting_height;
+    const PeerRef peer = GetPeerRef(nodeid);
+    if (peer == nullptr) {
+        return stats;
+    }
+    stats.m_services = peer->m_their_services;
+    if (peer->m_starting_height != -1) {
+        stats.m_starting_height = peer->m_starting_height;
+    }
     // It is common for nodes with good ping times to suddenly become lagged,
     // due to a new block arriving or other large transfer.
     // Merely reporting pingtime might fool the caller into thinking the node was still responsive,
     // since pingtime does not update until the ping is complete, which might take a while.
     // So, if a ping is taking an unusually long time in flight,
     // the caller can immediately detect that this is happening.
-    auto ping_wait{0us};
     if ((0 != peer->m_ping_nonce_sent) && (0 != peer->m_ping_start.load().count())) {
-        ping_wait = GetTime<std::chrono::microseconds>() - peer->m_ping_start.load();
+        stats.m_ping_wait = GetTime<std::chrono::microseconds>() - peer->m_ping_start.load();
     }
```

Comment:
SGTM

Reply:
Actually, we are only interested in non-zero values for pingwait, so it would be simpler not to do this. What do you think?

Reply:
Reverted this change in the latest push at b5255ca75f.

Reply:
Since the trend in this PR is to use `std::optional` it is odd to use the magic value `0` for `m_ping_wait`. Consider this change, which I think should be possible to be amended in bbd680d23e `p2p, rpc, qt: always provide CNodeStateStats, update getpeerinfo and gui` (not tested):

```diff
diff --git i/src/net_processing.h w/src/net_processing.h
index 2d7068f4c8..50fbe679a2 100644
--- i/src/net_processing.h
+++ w/src/net_processing.h
@@ -26,13 +26,13 @@ static const bool DEFAULT_PEERBLOCKFILTERS = false;
 static const int DISCOURAGEMENT_THRESHOLD{100};

 struct CNodeStateStats {
     std::optional<int> m_sync_height;
     std::optional<int> m_common_height;
     std::optional<int> m_starting_height;
-    std::chrono::microseconds m_ping_wait{0us};
+    std::optional<std::chrono::microseconds> m_ping_wait;
     std::vector<int> m_height_in_flight;
     std::optional<bool> m_relay_txs{false};
     CAmount m_fee_filter_received{0};
     uint64_t m_addr_processed = 0;
     uint64_t m_addr_rate_limited = 0;
     bool m_addr_relay_enabled{false};
diff --git i/src/qt/guiutil.cpp w/src/qt/guiutil.cpp
index 10e1ea169a..6562308052 100644
--- i/src/qt/guiutil.cpp
+++ w/src/qt/guiutil.cpp
@@ -750,13 +750,13 @@ QString formatServicesStr(quint64 mask)
     else
         return QObject::tr("Unknown");
 }

 QString formatPingTime(std::chrono::microseconds ping_time)
 {
-    return (ping_time == std::chrono::microseconds::max() || ping_time == 0us) ?
+    return ping_time == std::chrono::microseconds::max() ?
         QObject::tr("N/A") :
         QObject::tr("%1 ms").arg(QString::number((int)(count_microseconds(ping_time) / 1000), 10));
 }

 QString formatTimeOffset(int64_t nTimeOffset)
 {
diff --git i/src/qt/rpcconsole.cpp w/src/qt/rpcconsole.cpp
index 49402b0264..088608397f 100644
--- i/src/qt/rpcconsole.cpp
+++ w/src/qt/rpcconsole.cpp
@@ -1194,13 +1194,13 @@ void RPCConsole::updateDetailWidget()
     }
     ui->peerMappedAS->setText(stats->nodeStats.m_mapped_as != 0 ? QString::number(stats->nodeStats.m_mapped_as) : ts.na);
     ui->peerServices->setText(stats->nodeStateStats.m_services.has_value() ? GUIUtil::formatServicesStr(stats->nodeStateStats.m_services.value()) : ts.unknown);
     ui->peerSyncHeight->setText(stats->nodeStateStats.m_sync_height.has_value() ? QString("%1").arg(stats->nodeStateStats.m_sync_height.value()) : ts.unknown);
     ui->peerCommonHeight->setText(stats->nodeStateStats.m_common_height.has_value() ? QString("%1").arg(stats->nodeStateStats.m_common_height.value()) : ts.unknown);
     ui->peerHeight->setText(stats->nodeStateStats.m_starting_height.has_value() ? QString::number(stats->nodeStateStats.m_starting_height.value()) : ts.unknown);
-    ui->peerPingWait->setText(GUIUtil::formatPingTime(stats->nodeStateStats.m_ping_wait));
+    ui->peerPingWait->setText(stats->nodeStateStats.m_ping_wait.has_value() ? GUIUtil::formatPingTime(stats->nodeStateStats.m_ping_wait.value()) : QObject::tr("N/A"));
     ui->peerAddrRelayEnabled->setText(stats->nodeStateStats.m_addr_relay_enabled ? ts.yes : ts.no);
     ui->peerAddrProcessed->setText(QString::number(stats->nodeStateStats.m_addr_processed));
     ui->peerAddrRateLimited->setText(QString::number(stats->nodeStateStats.m_addr_rate_limited));
     ui->peerRelayTxes->setText(stats->nodeStateStats.m_relay_txs ? ts.yes : ts.no);
     ui->peersTabRightPanel->show();
 }
diff --git i/src/rpc/net.cpp w/src/rpc/net.cpp
index c688bd658f..bbfd88980d 100644
--- i/src/rpc/net.cpp
+++ w/src/rpc/net.cpp
@@ -213,14 +213,14 @@ static RPCHelpMan getpeerinfo()
         if (stats.m_last_ping_time > 0us) {
             obj.pushKV("pingtime", Ticks<SecondsDouble>(stats.m_last_ping_time));
         }
         if (stats.m_min_ping_time < std::chrono::microseconds::max()) {
             obj.pushKV("minping", Ticks<SecondsDouble>(stats.m_min_ping_time));
         }
-        if (statestats.m_ping_wait > 0s) {
-            obj.pushKV("pingwait", Ticks<SecondsDouble>(statestats.m_ping_wait));
+        if (statestats.m_ping_wait.has_value()) {
+            obj.pushKV("pingwait", Ticks<SecondsDouble>(statestats.m_ping_wait.value()));
         }
         obj.pushKV("version", stats.nVersion);
         // Use the sanitized form of subver here, to avoid tricksy remote peers from
         // corrupting or modifying the JSON output by putting special characters in
         // their ver message.
         obj.pushKV("subver", stats.cleanSubVer);
```